File: java/src/beecrypt/security/DSAPrivateKeyImpl.java
package beecrypt.security;

import java.io.*;
import java.math.BigInteger;
import java.security.*;
import java.security.interfaces.*;
import java.security.spec.*;

import beecrypt.io.*;

public class DSAPrivateKeyImpl implements DSAPrivateKey, Cloneable {
	private DSAParameterSpec _params;
	private BigInteger _x;
	private byte[] _enc;

	public DSAPrivateKeyImpl(BigInteger p, BigInteger q, BigInteger g,
			BigInteger x) {
		_params = new DSAParameterSpec(p, q, g);
		_x = x;
	}

	public DSAPrivateKeyImpl(DSAParameterSpec params, BigInteger x) {
		_params = params;
		_x = x;
	}

	public DSAPrivateKeyImpl(DSAPrivateKeySpec spec) {
		_params = new DSAParameterSpec(spec.getP(), spec.getQ(), spec.getG());
		_x = spec.getX();
	}

	public DSAPrivateKeyImpl(DSAPrivateKey key) {
		_params = new DSAParameterSpec(key.getParams().getP(), key.getParams()
				.getQ(), key.getParams().getG());
		_x = key.getX();
	}

	public Object clone() {
		return new DSAPrivateKeyImpl(this);
	}

	public boolean equals(Object obj) {
		if (this == obj)
			return true;

		if (obj instanceof DSAPrivateKey) {
			DSAPrivateKey pri = (DSAPrivateKey) obj;

			if (!pri.getParams().getP().equals(_params.getP()))
				return false;
			if (!pri.getParams().getQ().equals(_params.getQ()))
				return false;
			if (!pri.getParams().getG().equals(_params.getG()))
				return false;
			if (!pri.getX().equals(_x))
				return false;

			return true;
		}
		return false;
	}

	public DSAParams getParams() {
		return _params;
	}

	public BigInteger getX() {
		return _x;
	}

	public byte[] getEncoded() {
		if (_enc == null) {
			try {
				ByteArrayOutputStream bos = new ByteArrayOutputStream();
				BeeOutputStream bee = new BeeOutputStream(bos);

				bee.writeBigInteger(_params.getP());
				bee.writeBigInteger(_params.getQ());
				bee.writeBigInteger(_params.getG());
				bee.writeBigInteger(_x);
				bee.close();

				_enc = bos.toByteArray();
			} catch (IOException e) {
				// shouldn't occur
			}
		}

		return _enc;
	}

	public String getAlgorithm() {
		return "DSA";
	}

	public String getFormat() {
		return "BEE";
	}
}

File: c++/provider/DSAPrivateKeyImpl.cxx
/*
 * Copyright (c) 2004 Beeyond Software Holding BV
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include "beecrypt/c++/provider/DSAPrivateKeyImpl.h"
#include "beecrypt/c++/io/ByteArrayOutputStream.h"
using beecrypt::io::ByteArrayOutputStream;
#include "beecrypt/c++/beeyond/BeeOutputStream.h"
using beecrypt::beeyond::BeeOutputStream;

using namespace beecrypt::provider;

namespace {
	const String FORMAT_BEE("BEE");
	const String ALGORITHM_DSA("DSA");
}

DSAPrivateKeyImpl::DSAPrivateKeyImpl(const DSAPrivateKey& copy)
{
	_params = new DSAParameterSpec(copy.getParams());
	_x = copy.getX();
	_enc = 0;
}

DSAPrivateKeyImpl::DSAPrivateKeyImpl(const DSAPrivateKeyImpl& copy)
{
	_params = new DSAParameterSpec(*copy._params);
	_x = copy._x;
	_enc = 0;
}

DSAPrivateKeyImpl::DSAPrivateKeyImpl(const DSAParams& params, const BigInteger& x) : _x(x)
{
	_params = new DSAParameterSpec(params.getP(), params.getQ(), params.getG());
	_enc = 0;
}

DSAPrivateKeyImpl::DSAPrivateKeyImpl(const dsaparam& params, const mpnumber& x) : _x(x)
{
	_params = new DSAParameterSpec(BigInteger(params.p), BigInteger(params.q), BigInteger(params.g));
	_enc = 0;
}

DSAPrivateKeyImpl::DSAPrivateKeyImpl(const BigInteger& p, const BigInteger& q, const BigInteger& g, const BigInteger& x) : _x(x)
{
	_params = new DSAParameterSpec(p, q, g);
	_enc = 0;
}

DSAPrivateKeyImpl::~DSAPrivateKeyImpl()
{
	delete _params;
	delete _enc;
}

DSAPrivateKeyImpl* DSAPrivateKeyImpl::clone() const throw ()
{
	return new DSAPrivateKeyImpl(*this);
}

bool DSAPrivateKeyImpl::equals(const Object* obj) const throw ()
{
	if (this == obj)
		return true;

	if (obj)
	{
		const DSAPrivateKey* pri = dynamic_cast<const DSAPrivateKey*>(obj);
		if (pri)
		{
			if (pri->getParams().getP() != _params->getP())
				return false;

			if (pri->getParams().getQ() != _params->getQ())
				return false;

			if (pri->getParams().getG() != _params->getG())
				return false;

			if (pri->getX() != _x)
				return false;

			return true;
		}
	}

	return false;
}

const DSAParams& DSAPrivateKeyImpl::getParams() const throw ()
{
	return *_params;
}

const BigInteger& DSAPrivateKeyImpl::getX() const throw ()
{
	return _x;
}

const bytearray* DSAPrivateKeyImpl::getEncoded() const throw ()
{
	if (!_enc)
	{
		try
		{
			ByteArrayOutputStream bos;
			BeeOutputStream bee(bos);

			bee.writeBigInteger(_params->getP());
			bee.writeBigInteger(_params->getQ());
			bee.writeBigInteger(_params->getG());
			bee.writeBigInteger(_x);
			bee.close();

			_enc = bos.toByteArray();
		}
		catch (IOException&)
		{
		}
	}

	return _enc;
}

const String& DSAPrivateKeyImpl::getAlgorithm() const throw ()
{
	return ALGORITHM_DSA;
}

const String* DSAPrivateKeyImpl::getFormat() const throw ()
{
	return &FORMAT_BEE;
}
